src/ivector/agglomerative-bottom-up-clustering.cc

```diff
@@ -0,0 +1,196 @@
+    nclusters_ = npoints_ = uttlist.size();
+  }
+
+  void Cluster();
+
+ private:
+  BaseFloat ScoreLookup(int32 i, int32 j);
+  void Initialize();
+  void MergeClusters(int32 i, int32 j);
+
+  int32 ct_;
+  const std::vector<std::string> &uttlist_;
+  const std::unordered_map<std::string, BaseFloat> &score_map_;
+  BaseFloat max_dist_;
+  BaseFloat thresh_;
+  int32 min_clust_;
+  std::vector<int32> *assignments_;
+
+  std::vector<int32> tmp_assignments_;
+
+  std::unordered_map<std::pair<int32, int32>, BaseFloat,
+                     PairHasher<int32, int32>> cluster_score_map_;
+  std::unordered_map<int32, AHCCluster*> clusters_map_;
+  std::set<int32> active_clusters_;
+  int32 nclusters_;
+  int32 npoints_;
+  typedef std::pair<BaseFloat, std::pair<uint_smaller, uint_smaller> > QueueElement;
+  // Priority queue using greater (lowest distances are highest priority).
+  typedef std::priority_queue<QueueElement, std::vector<QueueElement>,
+      std::greater<QueueElement>  > QueueType;
+  QueueType queue_;
+};
+
+void AgglomerativeClusterer::Cluster() {
+  KALDI_VLOG(2) << "Initializing cluster assignments.";
+  Initialize();
+
+  KALDI_VLOG(2) << "Clustering...";
+  while (nclusters_ > min_clust_ && !queue_.empty()) {
+    std::pair<BaseFloat, std::pair<uint_smaller, uint_smaller> > pr = queue_.top();
+    int32 i = (int32) pr.second.first, j = (int32) pr.second.second;
+    queue_.pop();
+    if ((active_clusters_.find(i) != active_clusters_.end()) &&
+        (active_clusters_.find(j) != active_clusters_.end()))
+      MergeClusters(i, j);
+  }
+
+  std::vector<int32> new_assignments(npoints_);
+  int32 i = 0;
+  std::set<int32>::iterator it;
+  for (it = active_clusters_.begin(); it != active_clusters_.end(); ++it) {
+    ++i;
+    AHCCluster* cluster = clusters_map_[*it];
+    std::vector<int32>::iterator utt_it;
+    for (utt_it = cluster->utt_ids.begin();
+         utt_it != cluster->utt_ids.end(); ++utt_it)
+      new_assignments[*utt_it] = i;
+    delete(cluster);
+  }
+  assignments_->swap(new_assignments);
+}
+
+BaseFloat AgglomerativeClusterer::ScoreLookup(int32 i, int32 j) {
+  if (i < j)
+    return cluster_score_map_[std::make_pair(i, j)];
+  else
+    return cluster_score_map_[std::make_pair(j, i)];
+}
+
+void AgglomerativeClusterer::Initialize() {
+  KALDI_ASSERT(nclusters_ != 0);
+  for (int32 i = 0; i < nclusters_; i++) {
+    std::vector<int32> ids;
+    ids.push_back(i);
+    AHCCluster* c = new AHCCluster(++ct_, -1, -1, ids);
+    clusters_map_[ct_] = c;
+    active_clusters_.insert(ct_);
+
+    for (int32 j = i+1; j < nclusters_; j++) {
+      std::string utts_key;
+      if (uttlist_[i] < uttlist_[j])
+        utts_key = uttlist_[i]+uttlist_[j];
+      else
+        utts_key = uttlist_[j]+uttlist_[i];
+      BaseFloat score;
+      std::unordered_map<std::string, BaseFloat>::const_iterator it =
+          score_map_.find(utts_key);
+      if (it == score_map_.end())
+        score = max_dist_;
```

Comment:
The score is used for more than just queue purposes. It is also used in future distance calculations using those points. It would be possible to consider all cluster distances that include that pair to be "invalid" and just have a check when you merge two clusters on all of the new distances to compute on whether or not they contain a scoreless pair. I chose to do it this way for flexibility, especially since the "max_dist" parameter can be set to a smaller value that would not necessarily functionally invalidate all merges containing it.